File: src/leetcode/python/two_pointers/LinkedListCycleII_142.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class LinkedListCycleII_142(object):
    def detectCycle(self, head):
        if head is None:
            return None
        slow = head
        fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if slow == fast:
                break
        else:
            return None
        while head != slow:
            head = head.next
            slow = slow.next
        return head

File: src/leetcode/java/two_pointers/LinkedListCycleII_142.java
package leetcode.java.two_pointers;

public class LinkedListCycleII_142 {

  public ListNode detectCycle(ListNode head) {
    if (head == null) return null;
    ListNode slow = head;
    ListNode fast = head;
    boolean isCycle = false;
    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      slow = slow.next;
      if (fast == slow) {
        isCycle = true;
        break;
      }
    }
    if (!isCycle) return null;
    ListNode start = head;
    while (start != slow) {
      start = start.next;
      slow = slow.next;
    }
    return start;
  }

  class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
      val = x;
      next = null;
    }
  }
}
